File: randomizers/enemy_randomizer.py
import random
from typing import List
from models.enemy import EnemySpec

class EnemyRandomizer:
    def __init__(self, seed: int):
        self.rng = random.Random(seed)
        self.randomize_stats = True
        self.randomize_rewards = True
        self.shuffle_resistances = False
        self.scale_min = 0.8
        self.scale_max = 1.3

    def set_randomize_stats(self, value: bool) -> None:
        self.randomize_stats = value

    def set_randomize_rewards(self, value: bool) -> None:
        self.randomize_rewards = value

    def set_shuffle_resistances(self, value: bool) -> None:
        self.shuffle_resistances = value

    def set_scale_min(self, value: float) -> None:
        self.scale_min = value

    def set_scale_max(self, value: float) -> None:
        self.scale_max = value

    def process_enemies(self, enemies: List[EnemySpec]) -> None:
        for enemy in enemies:
            scale = self.scale_min + (self.rng.random() * (self.scale_max - self.scale_min))

            if self.randomize_stats:
                enemy.hp = max(1, min(99999, int(enemy.hp * scale)))
                enemy.mp = max(0, min(9999, int(enemy.mp * scale)))
                enemy.attack = max(1, min(9999, int(enemy.attack * scale)))
                enemy.defense = max(0, min(9999, int(enemy.defense * scale)))
                enemy.agility = max(0, min(9999, int(enemy.agility * scale)))

            if self.randomize_rewards:
                enemy.exp_reward = max(1, min(999999, int(enemy.exp_reward * scale)))
                enemy.fol_reward = max(0, min(999999, int(enemy.fol_reward * scale)))

            if self.shuffle_resistances:
                enemy.element_resist_mask = self.rng.randint(0, 255)

File: so1_randomizer_java_updated/com/serifsystemworks/sofd/randomizer/EnemyRandomizer.java
package com.serifsystemworks.sofd.randomizer;

import com.serifsystemworks.sofd.models.EnemySpec;
import java.util.List;
import java.util.Random;

public class EnemyRandomizer {

    private final Random rng;
    private boolean randomizeStats = true;
    private boolean randomizeRewards = true;
    private boolean shuffleResistances = false;
    private double scaleMin = 0.8;
    private double scaleMax = 1.3;

    public EnemyRandomizer(long seed) {
        this.rng = new Random(seed);
    }

    public void setRandomizeStats(boolean randomizeStats) {
        this.randomizeStats = randomizeStats;
    }

    public void setRandomizeRewards(boolean randomizeRewards) {
        this.randomizeRewards = randomizeRewards;
    }

    public void setShuffleResistances(boolean shuffleResistances) {
        this.shuffleResistances = shuffleResistances;
    }

    public void setScaleMin(double scaleMin) {
        this.scaleMin = scaleMin;
    }

    public void setScaleMax(double scaleMax) {
        this.scaleMax = scaleMax;
    }

    public void processEnemies(List<EnemySpec> enemies) {
        for (EnemySpec enemy : enemies) {
            double scale = scaleMin + (rng.nextDouble() * (scaleMax - scaleMin));

            if (randomizeStats) {
                enemy.setHp((int) Math.min(99999, Math.max(1, enemy.getHp() * scale)));
                enemy.setMp((int) Math.min(9999, Math.max(0, enemy.getMp() * scale)));
                enemy.setAttack((int) Math.min(9999, Math.max(1, enemy.getAttack() * scale)));
                enemy.setDefense((int) Math.min(9999, Math.max(0, enemy.getDefense() * scale)));
                enemy.setAgility((int) Math.min(9999, Math.max(0, enemy.getAgility() * scale)));
            }

            if (randomizeRewards) {
                enemy.setExpReward((int) Math.min(999999, Math.max(1, enemy.getExpReward() * scale)));
                enemy.setFolReward((int) Math.min(999999, Math.max(0, enemy.getFolReward() * scale)));
            }

            if (shuffleResistances) {
                enemy.setElementResistMask((byte) rng.nextInt(256));
            }
        }
    }
}